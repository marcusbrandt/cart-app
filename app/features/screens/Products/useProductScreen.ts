import { useEffect } from 'react';
import { useProductScreenProps } from './types';
import { useAppDispatch, useAppSelector } from '../../../store/hooks';
import { Product, Products } from '../../types';
import { addProducts, refreshProducts } from '../../../store/productsSlice';
import { addToCart, removeFromCart } from '../../../store/cartSlice';
import { fetchProducts } from '../../services/products.service';

const useProductScreen: () => useProductScreenProps = () => {
  const dispatch = useAppDispatch();

  const products = useAppSelector((state) => state.products as Product[]);
  const cart = useAppSelector((state) => state.cart as Product[]);

  useEffect(() => {
    dispatch(refreshProducts([]));
    fetchProducts()
      .then((response) => response.json())
      .then((data) =>
        (data as Products).products.map((product) => ({
          ...product,
          cart: false,
        })),
      )
      .then((data) => dispatch(addProducts(data)))
      .catch((error) => console.error(error));
  }, []);

  const buttonAdd = (id: number) => {
    const addProductCart = products.find((product) => product.id === id);
    if (addProductCart) {
      dispatch(addToCart(addProductCart));
    }
  };

  const buttonRemove = (id: number) => {
    dispatch(removeFromCart(id));
  };

  const productsWithCart = products.map((product) => {
    const productInCart = cart.find((item) => item.id === product.id);
    if (productInCart) {
      return { ...product, cart: true };
    }

    return { ...product, cart: false };
  });

  return {
    products: productsWithCart,
    buttonAdd,
    buttonRemove,
  };
};

export default useProductScreen;
