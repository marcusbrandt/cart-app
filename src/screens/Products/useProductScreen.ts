import {useEffect, useState} from "react";
import {Product, Products, useProductScreenT } from "./types";


const productsUrl = 'https://gist.githubusercontent.com/marcusbrandt/8507b74f6cc0247c9656a228cf0b3475/raw/92a63a70eec517932c9c6ebc26bec1dcdaf8621f/products.json';

const useProductScreen: () => useProductScreenT = () => {
    const [products, setProducts] = useState<Product[] | undefined>();

    useEffect(() => {
        const fetchProducts = async () => {
            return fetch(productsUrl)
        }
        fetchProducts()
            .then(response => response.json())
            .then(data => ((data as Products).products.map(product => ({...product, cart: false}))))
            .then(data => setProducts(data))
            .catch(error => console.error(error))
    }, [])

    const buttonAdd = (id: number) => {
        console.log('buttonAdd', id)
        const newProducts = products?.map(product => {
            if(product.id === id) {
                return {...product, cart: true}
            }
            return product
        })
        setProducts(newProducts)
    }

    const buttonRemove = (id: number) => {
        console.log('buttonRemove', id)
        const newProducts = products?.map(product => {
            if(product.id === id) {
                return {...product, cart: false}
            }
            return product
        })
        setProducts(newProducts)
    }

    return {
        products,
        buttonAdd,
        buttonRemove,
    }
};

export default useProductScreen;
