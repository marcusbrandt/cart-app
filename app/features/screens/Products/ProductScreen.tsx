import { ScrollView, View, Text } from 'react-native';
import { Container } from './styles';
import useProductScreen from './useProductScreen';
import Product from '../../components/Product';

export const ProductScreen = () => {
  const { products, buttonAdd, buttonRemove, isLoading, error } =
    useProductScreen();

  if (isLoading) {
    return (
      <View style={{ flex: 1, justifyContent: 'center' }}>
        <Text style={{ alignSelf: 'center' }}>Carregando...</Text>
      </View>
    );
  }

  if (error) {
    return (
      <View style={{ flex: 1, justifyContent: 'center' }}>
        <Text style={{ alignSelf: 'center' }}>
          Ocorreu um erro ao carregar os produtos
        </Text>
      </View>
    );
  }

  if (!products || products.length === 0) {
    return (
      <View style={{ flex: 1, justifyContent: 'center' }}>
        <Text style={{ alignSelf: 'center' }}>
          Não foi possível encontrar nenhum produto
        </Text>
      </View>
    );
  }

  return (
    <ScrollView style={{ flex: 1 }}>
      <Container>
        {products?.map((product) => (
          <Product
            key={product.id}
            product={product}
            buttonAdd={buttonAdd}
            buttonRemove={buttonRemove}
          />
        ))}
      </Container>
    </ScrollView>
  );
};
