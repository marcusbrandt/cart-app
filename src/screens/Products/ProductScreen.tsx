import Product from "@/src/components/Product";
import useProductScreen from "@/src/screens/Products/useProductScreen";
import {ScrollView} from "react-native";
import { Container } from './styles';

export const ProductScreen = () => {
    const {products, buttonAdd, buttonRemove} = useProductScreen();
    return (
        <ScrollView style={ { flex: 1 } }>
            <Container>
                {
                    products?.map(product => (
                        <Product
                            key={product.id}
                            product={product}
                            buttonAdd={buttonAdd}
                            buttonRemove={buttonRemove}
                        />))
                }
            </Container>
        </ScrollView>
    )
}
