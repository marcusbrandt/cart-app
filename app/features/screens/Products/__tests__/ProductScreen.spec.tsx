import React from 'react';
import { fireEvent } from '@testing-library/react-native';
import { ProductScreen } from '../ProductScreen';
import { render } from '../../../../../__tests__/__utils__/wrapperUtils';
import * as useProductScreen from '../useProductScreen';

const useProductScreenMock = jest.spyOn(useProductScreen, 'default');
const buttonRemove = jest.fn();
const buttonAdd = jest.fn();
const mockProducts = [
  {
    id: 1,
    name: 'Product 1',
    price: 10,
    image: 'https://via.placeholder.com/100',
    cart: false,
  },
  {
    id: 2,
    name: 'Product 2',
    price: 20,
    image: 'https://via.placeholder.com/100',
    cart: true,
  },
];

describe('ProductScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    useProductScreenMock.mockReturnValue({
      products: mockProducts,
      buttonRemove,
      buttonAdd,
      isLoading: false,
      error: undefined,
    });
  });
  it('renders empty product list message when no products are available', () => {
    useProductScreenMock.mockReturnValue({
      products: [],
      buttonAdd,
      buttonRemove,
      isLoading: false,
      error: undefined,
    });
    const { getByText } = render(<ProductScreen />);
    expect(getByText('Não foi possível encontrar nenhum produto')).toBeTruthy();
  });

  it('renders products in the product list', () => {
    const { getByText } = render(<ProductScreen />);
    expect(getByText('Product 1')).toBeTruthy();
    expect(getByText('Product 2')).toBeTruthy();
  });

  it('calls buttonAdd when a product is added to the cart', () => {
    const { getByTestId } = render(<ProductScreen />);
    fireEvent.press(getByTestId(`product-button-add-${mockProducts[0].id}`));
    expect(buttonAdd).toHaveBeenCalled();
  });

  it('calls buttonRemove when a product is added to the cart', () => {
    const { getByTestId } = render(<ProductScreen />);
    fireEvent.press(getByTestId(`product-button-remove-${mockProducts[1].id}`));
    expect(buttonRemove).toHaveBeenCalled();
  });
});
