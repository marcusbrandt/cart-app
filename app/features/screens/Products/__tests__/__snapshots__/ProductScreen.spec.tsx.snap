// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductScreen renders error message when an error occurs 1`] = `
<View
  style={
    {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      {
        "alignSelf": "center",
      }
    }
  >
    Ocorreu um erro ao carregar os produtos
  </Text>
</View>
`;

exports[`ProductScreen renders loading data message 1`] = `
<View
  style={
    {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      {
        "alignSelf": "center",
      }
    }
  >
    Carregando...
  </Text>
</View>
`;
